File: IAIso-v5.0/core/iaiso-java/iaiso-consent/src/main/java/io/iaiso/consent/Issuer.java
package io.iaiso.consent;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.security.PrivateKey;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.List;

/**
 * Mints signed consent tokens.
 *
 * <p>Use HS256 with a shared secret (the standard case) or RS256 with
 * a private key (for cross-service delegation).
 */
public final class Issuer {

    /** Supplier for the issuer's notion of "now" in seconds. */
    @FunctionalInterface
    public interface SecondsClock {
        long nowSeconds();
        static SecondsClock wallclock() {
            return () -> System.currentTimeMillis() / 1000L;
        }
    }

    private final byte[] hsKey;
    private final PrivateKey rsKey;
    private final Algorithm algorithm;
    private final String issuer;
    private final long defaultTtlSeconds;
    private final SecondsClock clock;
    private final SecureRandom rng = new SecureRandom();

    private Issuer(Builder b) {
        this.hsKey = b.hsKey;
        this.rsKey = b.rsKey;
        this.algorithm = b.algorithm;
        this.issuer = b.issuer;
        this.defaultTtlSeconds = b.defaultTtlSeconds;
        this.clock = b.clock != null ? b.clock : SecondsClock.wallclock();
    }

    public static Builder builder() {
        return new Builder();
    }

    /** Issue a token. */
    public Scope issue(String subject, List<String> scopes, String executionId,
                       Long ttlSeconds, JsonObject metadata) {
        long now = clock.nowSeconds();
        long ttl = ttlSeconds != null ? ttlSeconds : defaultTtlSeconds;
        long exp = now + ttl;
        String jti = generateJti();

        JsonObject claims = new JsonObject();
        claims.addProperty("iss", issuer);
        claims.addProperty("sub", subject);
        claims.addProperty("iat", now);
        claims.addProperty("exp", exp);
        claims.addProperty("jti", jti);
        JsonArray scopesArr = new JsonArray();
        for (String s : scopes) {
            scopesArr.add(s);
        }
        claims.add("scopes", scopesArr);
        if (executionId != null && !executionId.isEmpty()) {
            claims.addProperty("execution_id", executionId);
        }
        if (metadata != null && metadata.size() > 0) {
            claims.add("metadata", metadata);
        }

        String token = Jwt.sign(algorithm, claims, hsKey, rsKey);
        return new Scope(
            token, subject, new ArrayList<>(scopes),
            executionId, jti, now, exp,
            metadata != null ? metadata : new JsonObject());
    }

    /** Generate a 64-byte base64url-without-padding HS256 secret. */
    public static String generateHs256Secret() {
        byte[] buf = new byte[64];
        new SecureRandom().nextBytes(buf);
        return Jwt.b64url(buf);
    }

    private String generateJti() {
        byte[] buf = new byte[16];
        rng.nextBytes(buf);
        return HexFormat.of().formatHex(buf);
    }

    public static final class Builder {
        private byte[] hsKey;
        private PrivateKey rsKey;
        private Algorithm algorithm = Algorithm.HS256;
        private String issuer = "iaiso";
        private long defaultTtlSeconds = 3600;
        private SecondsClock clock;

        public Builder hsKey(byte[] v) { this.hsKey = v; return this; }
        public Builder rsKey(PrivateKey v) { this.rsKey = v; return this; }
        public Builder algorithm(Algorithm v) { this.algorithm = v; return this; }
        public Builder issuer(String v) { this.issuer = v; return this; }
        public Builder defaultTtlSeconds(long v) { this.defaultTtlSeconds = v; return this; }
        public Builder clock(SecondsClock v) { this.clock = v; return this; }

        public Issuer build() { return new Issuer(this); }
    }
}

File: IAIso-v5.0/core/iaiso-php/src/Consent/Issuer.php
<?php

declare(strict_types=1);

namespace IAIso\Consent;

/** Issues consent tokens. */
final class Issuer
{
    /**
     * @param string|null $hsKey raw HMAC bytes (HS256)
     * @param resource|\OpenSSLAsymmetricKey|string|null $rsKey OpenSSL private key (RS256)
     * @param callable():int $clock seconds-since-epoch clock
     */
    public function __construct(
        private readonly ?string $hsKey,
        private readonly mixed $rsKey,
        private readonly Algorithm $algorithm,
        private readonly string $issuer,
        private readonly int $defaultTtlSeconds,
        private $clock,
    ) {
        if (!is_callable($this->clock)) {
            throw new \InvalidArgumentException('clock must be callable returning int seconds');
        }
    }

    public static function builder(): IssuerBuilder
    {
        return new IssuerBuilder();
    }

    /**
     * Issue a fresh token.
     *
     * @param string[] $scopes
     * @param array<string,mixed>|null $metadata
     */
    public function issue(
        string $subject,
        array $scopes,
        ?string $executionId = null,
        ?int $ttlSeconds = null,
        ?array $metadata = null,
    ): Scope {
        $now = ($this->clock)();
        $ttl = $ttlSeconds ?? $this->defaultTtlSeconds;
        $exp = $now + $ttl;
        $jti = bin2hex(random_bytes(16));
        // Spec field order: iss, sub, iat, exp, jti, scopes, [execution_id], [metadata]
        $claims = [
            'iss' => $this->issuer,
            'sub' => $subject,
            'iat' => $now,
            'exp' => $exp,
            'jti' => $jti,
            'scopes' => $scopes,
        ];
        if ($executionId !== null) {
            $claims['execution_id'] = $executionId;
        }
        if ($metadata !== null && count($metadata) > 0) {
            $claims['metadata'] = $metadata;
        }
        $token = Jwt::sign($this->algorithm, $claims, $this->hsKey, $this->rsKey);
        return new Scope(
            token: $token,
            subject: $subject,
            scopes: $scopes,
            executionId: $executionId,
            jti: $jti,
            issuedAt: $now,
            expiresAt: $exp,
            metadata: $metadata ?? [],
        );
    }

    /** Generate a 64-byte base64url-no-pad HS256 secret. */
    public static function generateHs256Secret(): string
    {
        return Jwt::b64UrlEncode(random_bytes(64));
    }
}
